rust/quic/src/quic.rs

```diff
@@ -0,0 +1,263 @@
+r_declare_state_free!(r_quic_state_free,QuicParserState,{});
+
+#[derive(Debug,PartialEq)]
+pub struct QuicPacket<'a> {
+    pub flags_has_version: bool,
+    pub flags_has_reset: bool,
+    pub conn_fld_len: u8,
+    pub connection_id: Option<&'a[u8]>,
+    pub version: Option<&'a[u8]>,
+    pub seq_fld_len: u8,
+    pub seq_id: &'a[u8],
+    pub payload: &'a[u8],
+}
+
+named!(pub parse_quic<QuicPacket>,
+   do_parse!(
+       flags: bits!(tuple!(
+               take_bits!(u8, 2),   // reserved
+               take_bits!(u8, 2),   // seq
+               take_bits!(u8, 2),   // conn,
+               take_bits!(u8, 1),   // reset
+               take_bits!(u8, 1)))  // has ver
+
+       // get connection field length and take it
+       >> conn_len: value!(match flags.2 { 0x00 => 0, 0x01 => 1, 0x02 => 4, /* 0x03 */_ => 8, })
+       >> connid: cond!(flags.2 > 0, take!(conn_len))
+
+       // if version is present it's always 4 bytes
+       >> ver : cond!(flags.4 == 1, take!(4))
+
+       // sequence is 1, 2, 4 or 6 bytes
+       >> seq_len: value!(match flags.1 { 0x00 => 1, 0x10 => 2, 0x20 => 4, /* 0x30 */_ => 6, })
+       >> seqid: take!(seq_len)
+
+       >> pl: rest
+
+       >> (
+           QuicPacket {
+               flags_has_version:flags.4 == 1,
+               flags_has_reset:flags.3 == 1,
+               conn_fld_len:conn_len,
+               version:ver,
+               seq_fld_len:seq_len,
+               connection_id:connid,
+               seq_id:seqid,
+               payload: pl,
+           }
+   ))
+);
+
+struct QuicParser;
+
+impl<'a> RParser<QuicParserState<'a>> for QuicParser {
+    fn new_state() -> QuicParserState<'a> {
+        QuicParserState::new(b"blah")
+    }
+
+    fn probe(i: &[u8]) -> bool {
+        match parse_quic(i) {
+            IResult::Done(_, quic_packet) => {
+//                println!("rust::quic.rs::probe: match!");
+                true
+            }
+            IResult::Incomplete(_) => {
+                println!("rust: incomplete");
+                false
+            }
+            IResult::Error(_) => {
+                println!("rust: error");
+                false
+            }
+        }
+    }
+
+    fn parse(this: &mut QuicParserState, i: &[u8], direction: u8) -> u32 {
+        match parse_quic(i) {
+            IResult::Done(_, quic_packet) => {
+                //println!("quic.rs: {:?}", quic_packet);
+
+                let my_int = bytes_to_u64(quic_packet.seq_id); 
+                this.counter = this.counter + 1;
+//                println!("quic.rs: SEQ {:?} raw {:?}, msg {}", my_int, quic_packet.seq_id, this.counter);
+            }
+            IResult::Incomplete(_) => {
+                println!("rust: incomplete");
+            }
+            IResult::Error(_) => {
+                println!("rust: error");
+            }
+        }
+
+        //let d = parse_quic(i);
+        //println!("d {:?}", d);
+        let status = R_STATUS_OK;
```

Comment:
I think status should be set to OK at the beginning of the function, and or'ed with parser events, or set to error.
The case for incomplete is more difficult: if the protocols supports fragmentation, data must be copied in the parser state and tested in the next function call.